nixos/librechat: init module

rrvsh opened the pull request:
This is a NixOS module for [LibreChat](https://librechat.ai), a self-hostable enhanced ChatGPT clone with agents and RAG API capability, among others.

Closes #352044.

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [Nixpkgs 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/doc/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/doc/manual/release-notes/rl-2505.section.md) Nixpkgs Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
- [NixOS 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) NixOS Release notes)
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

rrvsh committed:
maintainers: add rrvsh

rrvsh committed:
nixos/librechat: init module

pluiedev commented on nixos/modules/services/web-apps/librechat.nix:
```diff
@@ -0,0 +1,149 @@
+      description = "Environment variables that will be set for the service. See https://www.librechat.ai/docs/configuration/dotenv for a full list.";
+    };
+    settings = lib.mkOption {
+      type = lib.types.attrs;
```
attrs is deprecated as it works poorly with options merging. Instead, RFC 42 recommends using a submodule with its freeform type set to the type accepted by the config file format (in this case, YAML)


```suggestion
      type = lib.types.submodule {
        # in the let-in: format = pkgs.formats.yaml { };
        freeformType = format.type;
      };
```

rrvsh replied:
Thanks for the heads up! I've amended all the attr options to use submodules instead.

pluiedev commented on nixos/modules/services/web-apps/librechat.nix:
```diff
@@ -0,0 +1,149 @@
+        User = cfg.user;
+        Group = cfg.group;
+        PermissionsStartOnly = "true"; # run mkdir as root
+        ExecStartPre = [
+          "${pkgs.coreutils}/bin/mkdir -p ${cfg.path}"
+          "${pkgs.coreutils}/bin/chown -R ${cfg.user}:${cfg.group} ${cfg.path}"
+          "${pkgs.coreutils}/bin/chmod 775 ${cfg.path}"
+        ];
```
Consider using something like `systemd.tmpfiles.settings` to make the system do this declaratively

rrvsh replied:
Wasn't aware of this, thanks. I've changed it to use tmpfiles

pluiedev commented on nixos/tests/librechat.nix:
```diff
@@ -0,0 +1,28 @@
+    services.librechat = {
+      enable = true;
+      env = {
+        CREDS_KEY = "6d6deb03cdfb27ea454f6b9ddd42494bdce4af25d50d8aee454ddce583690cc5";
+        CREDS_IV = "7c09a571f65ac793611685cc9ab1dbe7";
+        JWT_SECRET = "29c4dc7f7de15306accf5eddb4cb8a70eb233d9fba4301f8f47f14c8c047ac81";
+        JWT_REFRESH_SECRET = "f2c1685561f2f570b3e7955df267b5c602ee099f14dc5caa0dacc320580ea180";
+        MONGO_URI = "mongodb://localhost:27017";
```
Are these real credentials...?

rrvsh replied:
Regardless, I've added a comment explaining this in the test file to avoid any further confusion

pluiedev commented on nixos/modules/services/web-apps/librechat.nix:
```diff
@@ -0,0 +1,149 @@
+}:
+let
+  cfg = config.services.librechat;
+  configFile = pkgs.writeTextFile {
+    name = "librechat.yaml";
+    text = lib.generators.toYAML { } cfg.settings;
+  };
```
```suggestion
  format = pkgs.formats.yaml:
  configFile = format.generate "librechat.yaml" cfg.settings;
```

pluiedev commented on nixos/modules/services/web-apps/librechat.nix:
```diff
@@ -0,0 +1,149 @@
+  exportAllCredentials = vars: lib.concatStringsSep "\n" (lib.mapAttrsToList exportCredentials vars);
+  transformCredential = n: v: "${n}_FILE:${v}";
+  getLoadCredentialList = lib.mapAttrsToList transformCredential cfg.credentials;
+  allowedPorts = lib.optionals cfg.openFirewall (
+    if cfg.env ? PORT then [ cfg.env.port ] else [ 3080 ]
+  );
```
This doesn't look ideal. For vital configuration options like these we should offer a top level option with type `lib.types.port`

pluiedev reviewed (changes requested):
Welcome to Nixpkgs! Here's some suggestions to your changes

pluiedev commented on nixos/modules/services/web-apps/librechat.nix:
```diff
@@ -0,0 +1,162 @@
+      description = "The group to run the service as.";
+    };
+    credentials = lib.mkOption {
+      type =
+        with lib.types;
+        submodule {
+          freeformType = attrsOf path;
+        };
```
Is there any reason this should be a submodule? Just an `attrsOf path` makes far more sense IMO

pluiedev commented on nixos/modules/services/web-apps/librechat.nix:
```diff
@@ -0,0 +1,162 @@
+      })
+    ];
+    networking.firewall.allowedTCPPorts = lib.optionals cfg.openFirewall [ cfg.port ];
+    systemd.tmpfiles.rules = [
+      "d ${cfg.logDir} ${"0755"} ${cfg.user} ${cfg.group} -"
+    ];
```
It's just my personal opinion but IMO the systemd.tmpfiles.settings API looks better

```suggestion
    systemd.tmpfiles.settings."10-librechat"."${cfg.logDir}".d = {
      mode = "0755";
      inherit (cfg) user group;
    ];
```

pluiedev reviewed (changes requested):
Looks really really good save for a few points

rrvsh commented:
Made the requested changes. I'm a little unused to the idiomatic Nix way to do things, so I appreciate the guidance. Hope this diff works :)

pluiedev commented on nixos/modules/services/web-apps/librechat.nix:
```diff
@@ -0,0 +1,157 @@
+in
+{
+  options.services.librechat = {
+    enable = lib.mkEnableOption "Whether to enable the LibreChat server.";
```
mkEnableOption already prefixes the string with "Whether to enable", so this is effectively "Whether to enable Whether to enable..."

```suggestion
    enable = lib.mkEnableOption " LibreChat server.";
```

pluiedev commented on nixos/modules/services/web-apps/librechat.nix:
```diff
@@ -0,0 +1,157 @@
+    package = lib.mkPackageOption pkgs "librechat" { };
+    openFirewall = lib.mkEnableOption "Whether to open the port in the firewall.";
+    logDir = lib.mkOption {
+      type = lib.types.str;
```
```suggestion
      type = lib.types.path;
```

pluiedev commented on nixos/modules/services/web-apps/librechat.nix:
```diff
@@ -0,0 +1,157 @@
+  options.services.librechat = {
+    enable = lib.mkEnableOption "Whether to enable the LibreChat server.";
+    package = lib.mkPackageOption pkgs "librechat" { };
+    openFirewall = lib.mkEnableOption "Whether to open the port in the firewall.";
```
If a description would sound odd when prefixed with "Whether to enable...", you could also do this

```suggestion
    openFirewall = lib.mkEnableOption null // {
      description = "Whether to open the port in the firewall.";
    };
```

pluiedev commented on nixos/modules/services/web-apps/librechat.nix:
```diff
@@ -0,0 +1,157 @@
+
+  config = lib.mkIf cfg.enable {
+    assertions = [
+      (lib.mkIf (cfg.env ? PORT) {
+        assertion = cfg.port == cfg.env.PORT;
+        message = "You cannot set services.librechat.port and services.librechat.env.PORT to different values.";
+      })
```
I'm not even sure if what you wrote works, but this is how we usually do this (the `->` operator is a super cool way to say "given A, then B must also be true" and is used extensively in assertions like this)

```suggestion
      {
        assertion = cfg ? port -> cfg.port == cfg.env.PORT;
        message = "services.librechat.port and services.librechat.env.PORT must be set to equal values.";
      }
```

rrvsh replied:
oo this is really cool actually

pluiedev commented on nixos/modules/services/web-apps/librechat.nix:
```diff
@@ -0,0 +1,159 @@
+in
+{
+  options.services.librechat = {
+    enable = lib.mkEnableOption "the LibreChat server.";
```
There's already a period included also
```suggestion
    enable = lib.mkEnableOption "the LibreChat server";
```

rrvsh replied:
hahaha thank you for pointing this out - made the change

pluiedev reviewed (approved):
LGTM except for one tiny tiny tiny thing

adamcstephens commented:
There’s a merge conflict in the release notes.

rrvsh commented:
How do I fix that, exactly? Wouldn't this be something to fix upon a merge? Sorry, first time contributing to a large project

niklaskorz reviewed (commented):
This is a good start, but still needs some more work to get ready. We should check the for presence of credentials whose absence results in LibreChat crashing without any useful information (so the user knows what's going on). As mentioned I also need to refactor the log and storage paths in the package to use environment variables instead (and then open a PR on upstream, as this is generally a good idea to have regardless of NixOS).

I have uploaded my own config here, feel free to take what you need from it: https://gist.github.com/niklaskorz/01f3b57920555ccb6a4c39a6bebedc0c (the environment file sets `MONGO_URI`, `CREDS_KEY`, `CREDS_IV`, `MEILI_MASTER_KEY`, `JWT_SECRET`, `JWT_REFRESH_SECRET`, `OPENID_CLIENT_SECRET`, `OPENID_SESSION_SECRET`, `AZURE_API_KEY` and a few other vars)

rrvsh commented:
Force pushed to add the following changes:
- Add assertions to make sure users set the required environment variables for startup
- Fixed the merge conflict in rl notes
- Used systemd.services environment where possible
- added @niklaskorz as maintainer
- changed logDir to dataDir

rrvsh commented on nixos/modules/services/web-apps/librechat.nix:
```diff
@@ -0,0 +1,159 @@
+  };
+
+  meta.maintainers = [
+    lib.maintainers.rrvsh
```
done :+1:

rrvsh commented on nixos/modules/services/web-apps/librechat.nix:
```diff
@@ -0,0 +1,159 @@
+          export CONFIG_PATH=${configFile}
+
+          ${exportAll cfg.env}
+          ${exportAllCredentials cfg.credentials}
```
Security-wise, not really. This is an upstream issue IMO as everything is required to be set as environment variables. I'm primarily doing it this way so we can cat secret files without permission issues

rrvsh commented on nixos/modules/services/web-apps/librechat.nix:
```diff
@@ -0,0 +1,159 @@
+      };
+      script = # sh
+        ''
+          cd ${cfg.logDir}
```
I've renamed logDir to dataDir for now, but will wait out for the PR.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/5608

06kellyjac reviewed (changes requested):
Some observations.

---

Also just to doubly-double-check, is the code execution feature paid for and unlocked locally e.g. code executes locally? Or is it fully remote service where your code is executed by the librechat team?

https://www.librechat.ai/docs/features/code_interpreter#subscription

Based on the limitations section and the [why paid api section](https://www.librechat.ai/docs/features/code_interpreter#why-a-paid-api) I'm guessing execution is remote.

If it wasn't I'd want to be looking at how things are sandboxed running locally.

rrvsh commented:
I think it's technically possible to self host the code execution API, but only for their enterprise clients? The answer to your question is no though, the subscription just gates access to their hosted API. It doesn't run on your local machine. Thanks for the review!

rrvsh commented on nixos/modules/services/web-apps/librechat.nix:
```diff
@@ -0,0 +1,159 @@
+    };
+    systemd.services.librechat = {
+      wantedBy = [ "multi-user.target" ];
+      after = [ "tmpfiles.target" ];
```
What would be your thoughts on a design where each optional service is a submodule like the database, and each have createLocally, host port etc. attributes:

```nix
services.librechat.meilisearch = {
# Use this to set e.g. the service after and env vars
enable = true;
createLocally = true; # will set up the service
host = null;
port = null; 
};
# or
services.librechat.meilisearch = {
enable = true;
createLocally = false; # assert that host/port is not set if this is set or override the below if this is true
host = 192.168.0.2;
port = 1234; 
};
```

gepbird commented on nixos/tests/librechat.nix:
```diff
@@ -0,0 +1,30 @@
+  name = "librechat";
+
+  nodes.machine = {
+    # Run this test with NIXPKGS_ALLOW_UNFREE=1
```
This doesn't error out without `NIXPKGS_ALLOW_UNFREE=1` and automated tools like r-ryantm can run the tests on a package update.
```suggestion
  node.pkgsReadOnly = false;

  nodes.machine =
    { lib, ... }:
    {
      nixpkgs.config.allowUnfreePredicate =
        pkg:
        builtins.elem (lib.getName pkg) [
          "mongodb"
        ];
        
```

06kellyjac replied:
It does. It might be inheriting local config if youve allowed either all unfree or "mongodb" specifically.


```
λ gh pr checkout 412668
remote: Enumerating objects: 24, done.
remote: Counting objects: 100% (17/17), done.
remote: Compressing objects: 100% (2/2), done.
remote: Total 24 (delta 15), reused 15 (delta 15), pack-reused 7 (from 1)
Unpacking objects: 100% (24/24), 26.00 KiB | 682.00 KiB/s, done.
From github.com:nixos/nixpkgs
 * [new ref]                   refs/pull/412668/head -> librechat-module
Updating files: 100% (24598/24598), done.
Switched to branch 'librechat-module'
λ nix-build -A nixosTests.librechat
error:
       … while evaluating the attribute 'drvPath'
         at /home/jk/projects/personal/nixpkgs/lib/customisation.nix:418:7:
          417|     // {
          418|       drvPath =
             |       ^
          419|         assert condition;

       … while calling the 'derivationStrict' builtin
         at <nix/derivation-internal.nix>:37:12:
           36|
           37|   strict = derivationStrict drvAttrs;
             |            ^
           38|

       … while evaluating the option `nodes.machine.system.build.toplevel':

       … while evaluating definitions from `/home/jk/projects/personal/nixpkgs/nixos/modules/system/activation/top-level.nix':

       … while evaluating the option `nodes.machine.warnings':

       … while evaluating definitions from `/home/jk/projects/personal/nixpkgs/nixos/modules/system/boot/systemd.nix':

       … while evaluating the option `nodes.machine.systemd.services.mongodb.serviceConfig':

       … while evaluating definitions from `/home/jk/projects/personal/nixpkgs/nixos/modules/services/databases/mongodb.nix':

       (stack trace truncated; use '--show-trace' to show the full, detailed trace)

       error: Package ‘mongodb-7.0.16’ in /home/jk/projects/personal/nixpkgs/pkgs/servers/nosql/mongodb/mongodb.nix:186 has an unfree license (‘SSPL’), refusing to evaluate.

       a) To temporarily allow unfree packages, you can use an environment variable
          for a single invocation of the nix tools.

            $ export NIXPKGS_ALLOW_UNFREE=1

          Note: When using `nix shell`, `nix build`, `nix develop`, etc with a flake,
                then pass `--impure` in order to allow use of environment variables.

       b) For `nixos-rebuild` you can set
         { nixpkgs.config.allowUnfree = true; }
       in configuration.nix to override this.

       Alternatively you can configure a predicate to allow specific packages:
         { nixpkgs.config.allowUnfreePredicate = pkg: builtins.elem (lib.getName pkg) [
             "mongodb"
           ];
         }

       c) For `nix-env`, `nix-build`, `nix-shell` or any other Nix command you can add
         { allowUnfree = true; }
       to ~/.config/nixpkgs/config.nix.
```

If you run `nix build .#nixosTests.librechat` it should be more pure and force the error.

gepbird commented on nixos/modules/services/web-apps/librechat.nix:
```diff
@@ -0,0 +1,188 @@
+      };
+      default = { };
+      description = "Environment variables that will be set for the service. See https://www.librechat.ai/docs/configuration/dotenv for a full list.";
+    };
```
To put the port here (see previous comment), you can convert the `env` option to a submodule and move the `port` option declaration there, also capitalize it. Since its type should be kept as a port, but we will need a string later, we can use `lib.types.coercedTo` to make the conversion.

Currently after rendering the NixOS options documentation, the link in the description is not clickable, you can put it in a markdown link for example.

```suggestion
    env = lib.mkOption {
      type = lib.types.submodule {
        freeformType =
          with lib.types;
          attrsOf (oneOf [
            str
            path
            int
            port
            float
          ]);
        options = {
          PORT = lib.mkOption {
            type = with lib.types; coercedTo port toString str;
            default = 3080;
            example = 2309;
            description = "The value that will be passed to the PORT environment variable, telling LibreChat what to listen on.";
          };
        };
      };
      example = {
        ALLOW_REGISTRATION = true;
        HOST = "0.0.0.0";
        PORT = 2309;
        CONSOLE_JSON_STRING_LENGTH = 255;
      };
      description = ''
        Environment variables that will be set for the service.
        See [LibreChat environment variables](https://www.librechat.ai/docs/configuration/dotenv).
      '';
    };
```

gepbird commented on nixos/modules/services/web-apps/librechat.nix:
```diff
@@ -0,0 +1,188 @@
+      default = 3080;
+      example = 2309;
+      description = "The value that will be passed to the PORT environment variable, telling LibreChat what to listen on.";
+    };
```
```suggestion
```

06kellyjac commented:
I've been running this locally for a while and it's been going quite well.
We will need to adjust some of the librechat logic for uploads again because image upload is broken.
I've been trying to work on it in addition to updating to latest but I can't get `buildNpmPackage` to work with the update

rrvsh commented:
thanks for the reviews guys, ill get back to this in a couple of weeks, but feel free to pick up the PR in the meantime, ill close this one if you do. ive been using it locally since submitting this too fwiw and it has had no issues except when i fuck around with impermanence

gepbird commented:
I continued this in https://github.com/NixOS/nixpkgs/pull/462287, thanks @rrvsh for your work :)
